### PR TITLE
GIO endpoint

### PR DESCRIPTION
**Pull request by alexmikhalevich**
This PR implements a General Input/Output endpoint.

### CODE REVIEW

File: rollup-http/rollup-http-server/src/http_service.rs [hunk 1]
```diff
@@ -148,6 +150,25 @@ async fn report(report: Json<Report>, data: Data<Mutex<Context>>) -> HttpRespons
     };
 }
 
+/// Process gio request and return the result
+#[actix_web::post("/gio")]
+async fn gio(request: Json<GIORequest>, data: Data<Mutex<Context>>) -> HttpResponse {
+    log::debug!("received gio request {:#?}", request);
+    let context = data.lock().await;
+    // Write report to linux rollup device
```
Comment (mpolitzer) [inline rollup-http/rollup-http-server/src/http_service.rs]: I recognize I'm being pedantic with this comment, but here it goes:
This comment got inserted here and removed on the next commit that states it is only formatting the code.

Formatting first and then implementing a feature (while keeping the formatting) makes the review process easier, consider that for next time.

File: rollup-http/rollup-http-server/build.rs [hunk 2]
```diff
@@ -55,6 +55,7 @@ fn main() {
 
     // link the libcmt shared library
     println!("cargo:rustc-link-search=native={}", lib_dir);
+    println!("cargo:rerun-if-changed={}/libcmt.a", lib_dir);
```
Comment (mpolitzer) [inline rollup-http/rollup-http-server/build.rs:58]: nice catch!

File: sys-utils/libcmt/src/io-mock.c [hunk 3]
```diff
@@ -219,6 +219,14 @@ static int mock_tx_exception(cmt_io_driver_mock_t *me, struct cmt_io_yield *rr)
     return store_next_output(me, "exception-", &me->exception_seq, rr);
 }
 
+static int mock_tx_gio(cmt_io_driver_mock_t *me, struct cmt_io_yield *rr) {
```
Comment (mpolitzer) [inline sys-utils/libcmt/src/io-mock.c:222]: Do you think this is sufficient?

I was under the impression that we should implement a way to feed the GIO response back in.
I tried doing this in this draft: https://github.com/cartesi/machine-emulator-tools/pull/35.

File: sys-utils/libcmt/src/io-mock.c [hunk 4]
```diff
@@ -219,6 +219,14 @@ static int mock_tx_exception(cmt_io_driver_mock_t *me, struct cmt_io_yield *rr)
...
+    if (rr->cmd != HTIF_YIELD_CMD_MANUAL) {
+        (void) fprintf(stderr, "Expected cmd to be MANUAL\n");
+        return -EINVAL;
+    }
+    return store_next_output(me, "gio-", &me->exception_seq, rr);
```
Comment (mpolitzer) [inline sys-utils/libcmt/src/io-mock.c:227]: Add a separate sequence counter for gio requests.

**Review by mpolitzer — CHANGES_REQUESTED** [review]
Required changes:
- "Add a separate sequence counter for gio requests."

Rest left as comments

File: sys-utils/libcmt/src/io-mock.c (hunk 3 above)
Comment (mpolitzer) [reply sys-utils/libcmt/src/io-mock.c:219]: I think the library still should have something regardless of the Rust http server.
The GIO response doesn't necessarily have to be related to the request.
But to be able to test the response path, even the basic stuff.
A gio call should change values in `cmt_io_yield` and put something on rx buffer.

File: sys-utils/libcmt/src/io-mock.c [hunk 5]
```diff
@@ -117,6 +118,8 @@ static int load_next_input(cmt_io_driver_mock_t *me, struct cmt_io_yield *rr) {
     me->output_seq = 0;
     me->report_seq = 0;
     me->exception_seq = 0;
+    me->gio_seq = 0;
+    me->rx->end = me->rx->begin + file_length;
```
Comment (mpolitzer) [inline sys-utils/libcmt/src/io-mock.c:122]:
```suggestion
```

`me->tx` and `me->rx` represent the hardware memory regions mmaped to userspace and should be treated as constants.
The `file_length` information is already present in `rr->data` field.

Comment (mpolitzer) [reply sys-utils/libcmt/src/io-mock.c:122]: It makes sense.
I added a separate function to handle this `cmt_rollup_get_rx`.

https://github.com/cartesi/machine-emulator-tools/pull/29/files#diff-afdab1f6dcd5f7598ab6cd6d0a51a83922a24ccbd484e6eb4a63721e609d362cR222

Comment (mpolitzer) [reply sys-utils/libcmt/src/io-mock.c:122]: Marked this conversation as resolved by mistake. Please remove this line and use `cmt_rollup_get_rx` where appropriate.

**Review by mpolitzer — APPROVED** [review]

**Review by edubart — APPROVED** [review]

**Pull request merged**